src/protocols/rdp/channels/rdpdr/rdpdr-messages.c

```diff
@@ -0,0 +1,349 @@
+
+}
+
+/**
+ * Sends a Client Device List Announce Request message. The Client Device List
+ * Announce Request message is used by the client to enumerate all devices
+ * which should be made available within the RDP session via RDPDR. See:
+ *
+ * https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/10ef9ada-cba2-4384-ab60-7b6290ed4a9a
+ *
+ * @param svc
+ *     The guac_rdp_common_svc representing the static virtual channel being
+ *     used for RDPDR.
+ */
+static void guac_rdpdr_send_client_device_list_announce_request(guac_rdp_common_svc* svc) {
+
+    guac_rdpdr* rdpdr = (guac_rdpdr*) svc->data;
+
+    /* Calculate number of bytes needed for the stream */
+    int streamBytes = 16;
+    for (int i=0; i < rdpdr->devices_registered; i++)
+        streamBytes += rdpdr->devices[i].device_announce_len;
+
+    /* Allocate the stream */
+    wStream* output_stream = Stream_New(NULL, streamBytes);
+
+    /* Write header */
+    Stream_Write_UINT16(output_stream, RDPDR_CTYP_CORE);
+    Stream_Write_UINT16(output_stream, PAKID_CORE_DEVICELIST_ANNOUNCE);
+
+    /* Get the stream for each of the devices. */
+    Stream_Write_UINT32(output_stream, rdpdr->devices_registered);
+    for (int i=0; i<rdpdr->devices_registered; i++) {
+        
+        Stream_Write(output_stream,
+                Stream_Buffer(rdpdr->devices[i].device_announce),
+                rdpdr->devices[i].device_announce_len);
+
+        guac_client_log(svc->client, GUAC_LOG_DEBUG, "Registered device %i (%s)",
+                rdpdr->devices[i].device_id, rdpdr->devices[i].device_name);
+        
+    }
+
+    guac_rdp_common_svc_write(svc, output_stream);
+    guac_client_log(svc->client, GUAC_LOG_DEBUG, "All supported devices sent.");
+
+}
+
+void guac_rdpdr_process_server_announce(guac_rdp_common_svc* svc,
+        wStream* input_stream) {
+
+    unsigned int major, minor, client_id;
+
+    Stream_Read_UINT16(input_stream, major);
+    Stream_Read_UINT16(input_stream, minor);
+    Stream_Read_UINT32(input_stream, client_id);
+
+    /* Must choose own client ID if minor not >= 12 */
+    if (minor < 12)
+        client_id = random() & 0xFFFF;
+
+    guac_client_log(svc->client, GUAC_LOG_INFO, "Connected to RDPDR %u.%u as client 0x%04x", major, minor, client_id);
+
+    /* Respond to announce */
+    guac_rdpdr_send_client_announce_reply(svc, major, minor, client_id);
+
+    /* Name request */
+    guac_rdpdr_send_client_name_request(svc, ((guac_rdp_client*) svc->client->data)->settings->client_name);
+
+}
+
+void guac_rdpdr_process_clientid_confirm(guac_rdp_common_svc* svc,
+        wStream* input_stream) {
+    guac_client_log(svc->client, GUAC_LOG_DEBUG, "Client ID confirmed");
+}
+
+void guac_rdpdr_process_device_reply(guac_rdp_common_svc* svc,
+        wStream* input_stream) {
+
+    guac_rdpdr* rdpdr = (guac_rdpdr*) svc->data;
+
+    unsigned int device_id, ntstatus;
+    int severity, c, n, facility, code;
```

Comment:
Weirdly enough, they are standard names. The standard is not set by FreeRDP but by Microsoft. Each of these variables refers to a portion of a 32-bit "NTSTATUS" code:

https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/87fba13e-bf06-450e-83b1-9241dc81e781

As I'm calling the "sev" field "severity" here, it might make sense to change `c` to `customer`, however I'm not sure anyone knows what `n` is:

> **N (1 bit)**: Reserved. MUST be set to 0 so that it is possible to map an NTSTATUS value to an equivalent HRESULT value, as specified in section 2.1, by setting this bit.